cruise.umple/src/UmpleInternalParser_CodeMixset.ump

```diff
@@ -0,0 +1,178 @@
+/*
+
+Copyright: All contributers to the Umple Project
+
+This file is made available subject to the open source license found at:
+http://umple.org/license
+
+This file analyze core umple meta model tokens such as which language to generate,
+the current namespace, etc and populates the umple meta model.
+
+Please refer to UmpleInternalParser.ump for more details.
+
+*/
+
+
+namespace cruise.umple.compiler;
+
+class UmpleModel{
+  depend java.util.stream.*;
+
+/*
+ * returns a mixset from UmpleModel based on its name.
+ * @Abdulaziz   
+ */
+ Mixset getMixset(String name) {
+	Mixset mixset = null;
+	for(MixsetOrFile mOrF : getMixsetsOrFiles(name))
+	{
+		if (mOrF instanceof Mixset)
+		mixset = (Mixset)mOrF;
+	}
+	return mixset;
+ }
+/*
+ * returns UmpleFile from UmpleModel based on its name. 
+ * usage: to know if the file in which a mixset is mentioned was already used before. 
+ * @Abdulaziz   
+ */	
+ 
+ UmpleFile getUmpleFile(String name) {		
+ 	UmpleFile uFile = null;
+	for(MixsetOrFile mOrF : getMixsetsOrFiles(name))
+	{
+		if (mOrF instanceof UmpleFile)
+			uFile = (UmpleFile)mOrF;
+	}
+	return uFile;
+ }
+ 
+ List<MixsetOrFile> getMixsetsOrFiles(String name) {	 	
+ 	return mixsetOrFiles.stream().filter(mOrF -> mOrF.getName().equals(name)).collect(Collectors.toList());
+ }
+	
+}
+
+class UmpleInternalParser
+{
+	depend cruise.umple.compiler.UmpleFile;
+//	depend java.lang.StackTraceElement;
+//	depend java.lang.Thread;
+	 
+ // prepare mixsets that are inside an Umple class
+  private void analyzeMixsetDefinition(Token aToken , String className)
+  {
+  	aToken.addSubToken(new Token("entityName",className) );
+	aToken.addSubToken(new Token("entityType","class") );
+	analyzeMixset(aToken);		
+	}
+
+/*
+ * last action in the parse.	
+ */	
+ private void parseMixset(){
+ 	for (Mixset mixsetUse : model.getPossibleMixsetList())
```

Comment:
This loop seems excessively inefficient. All that we need to do is to process each element of the mixsetOrFile Association and it it is a mixset and their are fragments not already used, then process it. What you are doing is making a list and then searching the model for an element you have already found.